### DISCUSSION TITLE
Is it possible to provide pre-compiled parsers? (Windows users will benefit from it)

### DISCUSSION THREAD

**Discussion by jarmos-san** [nvim-treesitter/nvim-treesitter#1594] [General]
`nvim-treesitter` is amazing, so thanks for the great work!

That said, I figured there're plans to integrate syntax highlighting for most languages out-of-the-box ([source](https://www.reddit.com/r/neovim/comments/l7swp8/will_neovim_05_include_treesitter_languages_and/gl96vcf?utm_source=share&utm_medium=web2x&context=3) & the fact that a C parser is already being packaged with the `neovim` binary). So, I was assuming if it was possible to provide pre-compiled parsers in a straight-forward way? The biggest group of benefactors will be Windows-users if pre-compiled parsers were provided in a straightforward manner.

The current procedure is to either compile them our self on our local machines. Or download them as artifacts from the CI which isn't always possible as some tests either fail or the artifacts get expired. While compiling them on Windows isn't a big deal to be fair & I don't want to sound like I'm whining but it would be really nice if I didn't have to install extra tooling just for Neovim.

So, I look forward to discussing what would be the best way to distribute these parsers. Cheers!

**Comment by theHamsta** [comment]
Yes, we indirectly provide precompiled parsers but we have no mechanism to automatically download them at the moment. For each commit our CI compiles the parser libraries for all major platforms: https://github.com/nvim-treesitter/nvim-treesitter/actions/runs/1044691227 Just click on the tick next to a commit. For Windows, you should have 64bit binary to use the 64bit parsers. So users reported that only some of the artifacts from some of the compilers would work for them.

Ideally, there will be an automated way to get the parser binary directly from the parser when tree-sitter 1.0 gets released https://github.com/tree-sitter/tree-sitter/issues/930. So it would be could cool to have a shared mechanism for installation from which all tree-sitter projects can profit. So we would just download either a source tar or a pre-compiled version from a language/OS/architecture triplet. So I would prefer this solution compared to an automated download from CI.

**Reply by theHamsta** [reply]
Including parsers in the Neovim distribution is probably not a good idea.
- parsers a large (80MB are the artifacts compressed)
- updates can not be distributed easily but they need to be distributed together with the query configuration

It's probably best to download them on demand like currently the spell files or Atom language packs.

**Reply by jarmos-san** [reply]
> Including parsers in the Neovim distribution is probably not a good idea.

That's not what I meant & I wouldn't want the parsers distributed with the Neovim package either.

Rather is it possible for the parsers to be made available for download from the `nvim-treesitter/nvim-treesitter` repository, say through release assets?

Like right now, I can't find any pre-compiled parsers because the tests have either failed or the artifacts have expired.

I could compile them myself locally on my Windows machine but that would mean downloading MSVC or other similar tools. And if possible I would prefer not to need those tools just for compiling Neovim stuff.

So, from my perspective it makes sense to make pre-compiled parsers available for download right here in this repo. I mean, there's a CI & other automation toolsets ready to-use for automatically building & compiling the parsers. All that's left is to deliver the assets in a way that's quite straightforward.

**Reply by theHamsta** [reply]
Maybe, I think we could provide both solutions relatively easily. It would just be a great relief for use to remove all the installation logic since the Tree-Sitter project already plans to a have unified distribution strategy for parsers.

I thought that we could then just generate a download link for each configuration and that curl or curl+install.sh would be the only thing left for our plug-in.

Otherwise, we would have to upload a release for each parser and OS so that we would have a download link per parser. I think this could be done in the compilation workflow. We didn't do that yet because this can get complicated really fast FreeBSD/MacOS/Win32/win64 x amd64/arm64 . Source distribution was just an easier option and binary distribution would be easier as a shared effort together with the tree-sitter maintainers. emacs-tree-sitter could use that option, too. They are right now distributing binary packages only.

**Reply by theHamsta** [reply]
But PRs are welcome that upload the compiled parsers to the release page. I don't know whether there is a storage limit because in theory we would need to store each revision of all parsers we have released
